### PR TITLE
Event Manager / Propensity Client

### PR DESCRIPTION
**Pull request by mborof**
This is the bulk of changes required by Propensity to start using the events manager.  Still to do would be actually sending the new data to the server (and logging it).  There will be a future branch focusing on just that.

### CODE REVIEW

**Comment by mborof** [comment]
Hold off until the other event manager branch is merged

**Comment by mborof** [comment]
This is ready for review.

File: src/runtime/propensity.js [hunk 1]
```diff
@@ -15,7 +15,23 @@
...
 import * as PropensityApi from '../api/propensity-api';
 import {PropensityServer} from './propensity-server';
-import {isObject} from '../utils/types';
+import {isObject,isEnumValue} from '../utils/types';
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+import {Event} from '../api/propensity-api';
+
+/** @private @const {!Object<string,AnalyticsEvent>} */
+const PropensityEventToAnalyticsEvent = {
+  [Event.IMPRESSION_PAYWALL]: AnalyticsEvent.IMPRESSION_PAYWALL,
```
Comment (sohanirao) [inline src/runtime/propensity.js]: why square brackets?

File: src/runtime/propensity-server.js [hunk 2]
```diff
@@ -14,7 +14,34 @@
...
+
+/** @private @const {!Object<number,string>} */
+const AnalyticsEventToPropensityEvent = {
+  [AnalyticsEvent.UNKNOWN]: null,
+  [AnalyticsEvent.IMPRESSION_PAYWALL]: Event.IMPRESSION_PAYWALL,
+  [AnalyticsEvent.IMPRESSION_AD]: Event.IMPRESSION_AD,
+  [AnalyticsEvent.IMPRESSION_OFFERS]: Event.IMPRESSION_OFFERS,
+  [AnalyticsEvent.IMPRESSION_SUBSCRIBE_BUTTON]: null,
+  [AnalyticsEvent.IMPRESSION_SMARTBOX]: null,
+  [AnalyticsEvent.ACTION_SUBSCRIBE]: null,
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: AnalytisEvent.ACTION_SUBSCRIBE maps to Event.ACTION_PAYMENT_FLOW_STARTED

File: src/runtime/propensity.js (hunk 1 above)
Comment (mborof) [reply src/runtime/propensity.js]: If you take out the square brackets the syntax doesn't compile.  I wasn't sure how to actually make the keys defined from an enum value until I saw it done with the [...] in another file.

File: src/runtime/propensity-server.js (hunk 2 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: There is also an AnalyticsEvent.ACTION_PAY_FLOW_STARTED.

Comment (mborof) [reply src/runtime/propensity-server.js]: I see that ACTION_SUBSCRIBE is used in 1 place in the codebase: it might be preferable to replace it with ACTION_PAYMENT_FLOW_STARTED if that is in fact what it means.  It's a much better enum value

File: src/runtime/experiment-flags.js [hunk 3]
```diff
@@ -43,12 +43,18 @@ export const ExperimentFlags = {
   CONTRIBUTIONS: 'contributions',
 
   /**
-   * Enables the Propensity feature
+   * Enables the Propensity feature.
```
Comment (sohanirao) [inline src/runtime/experiment-flags.js]: this can be removed.

File: src/runtime/propensity-test.js [hunk 4]
```diff
@@ -166,4 +172,60 @@ describes.realWin('Propensity', {}, env => {
       expect(propensityScore.body.result).to.equal(42);
     });
   });
+
+  it('should convert all propensity events to analytics events', () => {
+    let eventSent = null;
+    const params = /** @type {JsonObject} */ ({
+      'source': 'user-action',
```
Comment (sohanirao) [inline src/runtime/propensity-test.js]: lets remove is_active since we plan to clarify the API anyway.
And let's use a source that's more believable like  'email'

Test code can serve as an example of API usage so lets be careful, although I recognize it does not make any difference to the test logic.

File: src/runtime/propensity-test.js [hunk 5]
```diff
@@ -126,24 +128,28 @@ describes.realWin('Propensity', {}, env => {
...
+  it('should send events to event manager', () => {
     let eventSent = null;
-    let paramsSent = null;
-    const params = /** @type {JsonObject} */ ({'source': 'user-action'});
-    const eventParams = JSON.stringify(params);
-    const propensityEvent = {
-      name: PropensityApi.Event.IMPRESSION_OFFERS,
+    const params = /** @type {JsonObject} */ ({
+      'source': 'user-action',
+      'is_active': false,
```
Comment (sohanirao) [inline src/runtime/propensity-test.js]: lets remove is_active and the value for 'source' to email.

File: src/runtime/experiment-flags.js (hunk 3 above)
Comment (mborof) [reply src/runtime/experiment-flags.js]: ok

File: src/runtime/propensity-test.js (hunk 5 above)
Comment (mborof) [reply src/runtime/propensity-test.js]: ok

File: src/runtime/propensity-test.js (hunk 4 above)
Comment (mborof) [reply src/runtime/propensity-test.js]: Ok

File: src/runtime/propensity-server.js [hunk 6]
```diff
@@ -39,6 +67,12 @@ export class PropensityServer {
     this.xhr_ = new Xhr(win);
     /** @private @const {number} */
     this.version_ = 1;
+
+    eventManager.registerEventListener(event => this.sendClientEvent_(event));
+
+    /** @private @const {!boolean} */
+    this.logSwgEvents_ = isExperimentOn(win,
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: // TODO(mborof): b/133519525

File: src/runtime/propensity-server.js [hunk 7]
```diff
@@ -125,6 +159,28 @@ export class PropensityServer {
     return this.xhr_.fetch(url, init);
   }
 
+  /**
+   *
+   * @param {!../api/client-event-manager-api.ClientEvent} event
+   */
+  sendClientEvent_(event) {
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: I would call this handleClientEvents_() since we are not always sending events.

File: src/runtime/propensity-server.js [hunk 8]
```diff
@@ -116,7 +150,7 @@ export class PropensityServer {
     if (context) {
       eventInfo = eventInfo + ':' + encodeURIComponent(context);
     }
-    let url = ServiceUrl.adsUrl('/subopt/data?events=')
+    let url = adsUrl('/subopt/data?events=')
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: Line 143: Any reason this method should stay public?

if the only caller is from within we can rename this to sendEvent_() and make it private.

File: src/runtime/propensity.js [hunk 9]
```diff
@@ -69,18 +89,19 @@ export class Propensity {
 
   /** @override */
   sendEvent(userEvent) {
-    if (!Object.values(PropensityApi.Event).includes(userEvent.name)) {
+    if (!isEnumValue(PropensityApi.Event, userEvent.name)
+        || !PropensityEventToAnalyticsEvent[userEvent.name]) {
       throw new Error('Invalid user event provided');
```
Comment (sohanirao) [inline src/runtime/propensity.js]: Do you see any issues in augmenting the error message on line 94 to include the userEvent.name?

File: src/runtime/propensity-server.js (hunk 2 above)
Comment (sohanirao) [reply src/runtime/propensity-server.js]: my bad: This is the right mapping.
  [AnalyticsEvent.ACTION_PAYMENT_FLOW_STARTED]:
      Event.ACTION_PAYMENT_FLOW_STARTED,

ACTION_SUBSCRIBE tracks click on subscribe button. There's no equivalent of that in Propensity world. Please ignore this whole set.

File: src/runtime/propensity-server.js (hunk 6 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: Ok

File: src/runtime/propensity-server.js (hunk 7 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: Ok

File: src/runtime/propensity-server-test.js [hunk 10]
```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
           expect(() => {throw reason;}).to.throw(/Invalid request/);
         });
   });
+
+  it('should listen for events from event manager', function*() {
+    //event registration happens during initialization so we need to stub it
+    //and then create a new propensity server
+    const eventMan = new ClientEventManager();
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: Why not use eventManager object created in line 43?

File: src/runtime/propensity-server-test.js [hunk 11]
```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
...
+    let receivedEvent = null;
+    sandbox.stub(PropensityServer.prototype, 'sendClientEvent_', event => {
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: Line 321 and similar:
Instead of stubbing an internal method of the class under test, the right way to do is to mock the method of EventManager, registerEventListener and capture the argument to that call. and that verifies that PropensityServer listens for events from the event manager. None of the event manager specific code should be tested here.

A helper method that returns the event sent in the URL
/**
 * @param {string} capturedUrl
 * @return {?PropensityEvent}
 */
 getPropensityEventFromUrl(capturedUrl) {
        const queryString = capturedUrl.split('?')[1];
        const queries = parseQueryString(queryString);
        expect('events' in queries).to.be.true;

        const events = decodeURIComponent(queries['events'].split(':'));
       const eventName = events[1];
       const eventParam = parseJson(events[2]);
       const propensityEvent = /** @typedef {PropensityEvent} */({
         name: eventName,
         active: eventParam['is_active'],
         data: eventParam,
       } 
        expect(events).to.equal(JSON.stringify(eventParam));

The best way to test the mapping is to capture what gets sent in the fetch API.
    let capturedUrl;
    let capturedRequest;
    sandbox.stub(Xhr.prototype, 'fetch',
        (url, init) => {
          capturedUrl = url;
          capturedRequest = init;
          return Promise.reject('Server down');
        });

Then when you invoke the captured handler with different events, you can test your can test what was sent in the URL
return capturedListenerCallback(defaultAnalyticsEvent).then(() => {
  throw new Error('must have failed');
}).catch(() => {
        const propensityEventSent = this.getPropensityEventFromUrl(capturedUrl);
       // expectation on this event.
        expect(() => {
          throw reason;
        }).to.throw(/Server down/);
});

File: src/runtime/propensity-server.js [hunk 12]
```diff
@@ -14,7 +14,34 @@
...
 import {Xhr} from '../utils/xhr';
-import * as ServiceUrl from './services';
+import {adsUrl} from './services';
+import {Event} from '../api/propensity-api';
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+import {isObject,isBoolean} from '../utils/types';
+import {ExperimentFlags} from './experiment-flags';
+import {isExperimentOn} from './experiments';
+
+/** @private @const {!Object<number,string>} */
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: This is @private to which class? Should this be inside the constructor?

File: src/runtime/propensity-server-test.js [hunk 13]
```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
...
+    new PropensityServer(win, pubId, eventMan);
+    eventMan.logEvent(defaultEvent);
+    yield eventMan.lastAction_;
+    expect(defaultEvent === receivedEvent).to.be.true;
+  });
+
+  it('should convert analytics events to propensity events', () => {
+    let receivedType = null;
+    let receivedContext = null;
+    sandbox.stub(PropensityServer.prototype, 'sendEvent',
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: I see you are trying to test the mapping here. Since sendEvent_ is going to be a private method and nobody outside is calling it, testing arguments passed in between private methods inside a class amounts to testing implementation. We need to test behavior. 

Do you need to test the mapping?

File: src/runtime/propensity-server-test.js [hunk 14]
```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
...
+        case AnalyticsEvent.EVENT_CUSTOM:
+          expect(receivedType).to.equal(PropensityApi.Event.EVENT_CUSTOM);
+          break;
+        default:
+          expect(false).to.be.true; //add new analytics event to the switch
+      }
+    }
+  });
+
+  it('should respect the SwG logging experiment flag', () => {
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: Better to break this up into two tests.
// Mock the EventManager.registerEventListener() method and capture the callback

     numFetchApiCalls = 0;
    sandbox.stub(Xhr.prototype, 'fetch',
        (url, init) => {
          numFetchApiCalls++;
          return Promise.reject(new Error('Publisher not whitelisted'));
        });

One where it 'should send SwG events to Propensity Service'
// Enable experiment flag
// invoke the callback and catch
captureCallback(swgOriginatedEvent).then(() => {
  throw new Error('must have failed');
}).catch(reason => {
  expect(numFetchApiCalls).to.equal(0);
  expect(() => throw reason).to.throw(/Server down/);
});

One where it 'should not send SwG events to Propensity Service'
// same setup
captureCallback(swgOriginatedEvent).then(() => {
  throw new Error('must have failed');
}).catch(reason => {
  expect(numFetchApiCalls).to.equal(0);
  expect(() => throw reason).to.throw(/Server down/);
});

File: src/runtime/propensity-server-test.js [hunk 15]
```diff
@@ -296,4 +312,132 @@ describes.realWin('PropensityServer', {}, env => {
...
+    let receivedType = null;
+    let receivedContext = null;
+    sandbox.stub(PropensityServer.prototype, 'sendEvent',
+        (eventType, context) => {
+          receivedType = eventType;
+          receivedContext = context;
+        });
+
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: Better to avoid testing the internal implementation if there's another way.

File: src/runtime/propensity-server-test.js (hunk 10 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: The comment above it explains why I had done it.  I've actually changed the implementation so it might not be necessary now - it depends how I register the listener.  

The syntax now: eventManager.registerEventListener(event => this.handleClientEvent_(event));
The way I had registered it initially: eventManager.registerEventListener(this.handleClientEvent_.bind(this));

For the second implementation, if I had stubbed it after the registration occurred then it would remember the original version of the function rather than the stubbed version (so it didn't work).  The way I'm doing it now doesn't matter.  I can do it either way.  I'm not sure which method is preferable.

File: src/runtime/propensity-server.js (hunk 12 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: It's a non-exported private const used for mapping different event types.  I could certainly put it inside the class if you think it should go there.  It's used exclusively by 1 line of code in propensity-server.

File: src/runtime/propensity-server.js (hunk 8 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: Agreed

File: src/runtime/propensity.js (hunk 9 above)
Comment (mborof) [reply src/runtime/propensity.js]: I think its a good idea

File: src/runtime/propensity-server-test.js (hunk 13 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: I initially thought I didn't but then later thought it would be better if I did.  I'm not sure - do I need to test the mapping?    I lean towards yes in case someone accidentally changes it but I'm not really sure.

Comment (sohanirao) [reply src/runtime/propensity-server-test.js]: as discussed, lets create another class to centralize event mapping across different entities that use it.

File: src/runtime/propensity-server-test.js (hunk 11 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: Done!

File: src/runtime/propensity-server-test.js (hunk 13 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: Done!

File: src/runtime/propensity-server-test.js (hunk 14 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: Ok

File: src/runtime/propensity-server-test.js (hunk 15 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: I think this is done

File: src/runtime/propensity.js [hunk 16]
```diff
@@ -69,18 +75,35 @@ export class Propensity {
...
     }
-    // TODO(sohanirao, mborof): Idenfity the new interface with event
-    // manager and update the lines below to adhere to that interface
-    let paramString = null;
-    if (userEvent.data) {
-      paramString = JSON.stringify(userEvent.data);
+
+    //event manager requires data to be a real object so fill it in if
+    //the publisher didn't pass the field in
+    if (!userEvent.data) {
```
Comment (sohanirao) [inline src/runtime/propensity.js]: Line 86, 87 feels like redundant code

Comment (sohanirao) [reply src/runtime/propensity.js]: if (userEvent.data && !isObject(userEvent.data)) {
      throw new Error('Event data must be an Object');
    }

Any reason why this code was not sufficient?

File: src/runtime/propensity-server.js [hunk 17]
```diff
@@ -39,6 +48,16 @@ export class PropensityServer {
     this.xhr_ = new Xhr(win);
     /** @private @const {number} */
     this.version_ = 1;
+
+    eventManager.registerEventListener(event => this.handleClientEvent_(event));
+
+    // TODO(mborof): b/133519525
+    /** @private @const {!boolean} */
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: I think we can keep it as this.logSwgEvents_ = isExperimentOn(...)

and then and it with the enable logging flag. No?

Any reason we need to keep track of two flags?

File: src/runtime/propensity-server-test.js [hunk 18]
```diff
@@ -76,7 +114,7 @@ describes.realWin('PropensityServer', {}, env => {
           return Promise.reject(new Error('Not sent from allowed origin'));
         });
     const eventParam = {'is_active': false, 'offers_shown': ['a', 'b', 'c']};
-    return propensityServer.sendEvent(
+    return propensityServer.sendEvent_(
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: this should happen in the then of the callback invocation. first capture the callback from mocking the registerEventListener interface.

File: src/runtime/propensity-server-test.js (hunk 15 above)
Comment (sohanirao) [reply src/runtime/propensity-server-test.js]: you are still calling handleClientEvent_() which is a internal private method.

You need to capture the callback registered if you want to invoke the listener.

File: src/runtime/propensity-server-test.js [hunk 19]
```diff
@@ -296,4 +334,81 @@ describes.realWin('PropensityServer', {}, env => {
...
+      return Promise.reject('Server down');
+    });
+
+    setExperiment(win, ExperimentFlags.LOG_SWG_TO_PROPENSITY, true);
+    eventManagerMock.expects('registerEventListener').once();
+    propensityServer = new PropensityServer(win, pubId, eventManager);
+
+    //both experiment and enable: ensure it actually logs
+    propensityServer.enableLoggingGoogleEvents();
+    propensityServer.handleClientEvent_(defaultEvent);
```
Comment (sohanirao) [inline src/runtime/propensity-server-test.js]: should not have to call internal private methods to do this test.

File: src/runtime/propensity.js (hunk 16 above)
Comment (mborof) [reply src/runtime/propensity.js]: It caused the test code to fail so I figured it was supposed to work

Comment (mborof) [reply src/runtime/propensity.js]: It didn't fail before because nothing had a hard requirement that this was an object until event manager

Comment (sohanirao) [reply src/runtime/propensity.js]: Here's a piece of code that tested it already
it('should provide valid event', () => {
    const correctEvent = {
      name: PropensityApi.Event.IMPRESSION_PAYWALL,
      active: false,
    };
    expect(() => {
      propensity.sendEvent(correctEvent);
    }).to.not.throw('Invalid user event provided');
    const incorrectEvent = {
      name: 'user-redirect',
    };
    expect(() => {
      propensity.sendEvent(incorrectEvent);
    }).to.throw('Invalid user event provided');
    const incorrectEventParam = {
      name: PropensityApi.Event.IMPRESSION_OFFERS,
      active: true,
      data: 'all_offers',
    };
    expect(() => {
      propensity.sendEvent(incorrectEventParam);
    }).to.throw(/Event data must be an Object/);
  });

Still not sure how your code is different other than splitting up the checks into two lines?

File: src/runtime/propensity-test.js [hunk 20]
```diff
@@ -70,32 +78,10 @@ describes.realWin('Propensity', {}, env => {
     }).throw(/Entitlements must be an Object/);
   });
 
-
-  it('should provide valid event', () => {
```
Comment (sohanirao) [inline src/runtime/propensity-test.js]: why is this test removed?

Comment (mborof) [reply src/runtime/propensity-test.js]: I reordered the tests to group logical items together.  It was just moved.

File: src/runtime/propensity.js (hunk 16 above)
Comment (mborof) [reply src/runtime/propensity.js]: Here is the specific test that used to work that would no longer work if I throw an error without a data object:

    const correctEvent = {
      name: PropensityApi.Event.IMPRESSION_PAYWALL,
      active: false,
    };
    expect(() => {
      propensity.sendEvent(correctEvent);
    }).to.not.throw('Invalid user event provided');

File: src/runtime/propensity-server-test.js (hunk 18 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: I think I did what you wanted

File: src/runtime/propensity-server-test.js (hunk 15 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: Now I think I know what you mean.  Done?

File: src/runtime/propensity-server-test.js (hunk 19 above)
Comment (mborof) [reply src/runtime/propensity-server-test.js]: Done?

File: src/runtime/propensity-server-test.js (hunk 15 above)
Comment (sohanirao) [reply src/runtime/propensity-server-test.js]: just realized one more thing. since you are programming is_active field into the event parameters based on active field of the PropensityEvent, it's better to test that as well.

Comment (mborof) [reply src/runtime/propensity-server-test.js]: Done

File: src/runtime/propensity-server.js [hunk 21]
```diff
@@ -201,4 +246,8 @@ export class PropensityServer {
           return this.parsePropensityResponse_(response);
         });
   }
+
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: Since this is a public API needs documentation.

File: src/runtime/propensity-server.js [hunk 22]
```diff
@@ -39,6 +48,16 @@ export class PropensityServer {
     this.xhr_ = new Xhr(win);
     /** @private @const {number} */
     this.version_ = 1;
+
+    eventManager.registerEventListener(event => this.handleClientEvent_(event));
```
Comment (sohanirao) [inline src/runtime/propensity-server.js]: We need to bind the callback to this instance of the class. 

eventManager.registerEventListener(this.handleClientEvent_.bind(this));

File: src/runtime/propensity-server.js (hunk 21 above)
Comment (sohanirao) [reply src/runtime/propensity-server.js]: And lets rename to enableLoggingSwgEvents()

These are not google events.

Comment (mborof) [reply src/runtime/propensity-server.js]: Done

File: src/runtime/propensity-server.js (hunk 22 above)
Comment (mborof) [reply src/runtime/propensity-server.js]: It seemed to work either way, I wasn't sure which was preferable.  I've changed it back to a bind

File: src/runtime/propensity.js [hunk 23]
```diff
@@ -69,27 +76,44 @@ export class Propensity {
...
     }
-    if (userEvent.data) {
-      paramString = JSON.stringify(userEvent.data);
+
+    if (isBoolean(userEvent.active)) {
+      if (!data) {
+        data = {};
+      }
+      Object.assign(data, {'is_active': userEvent.active});
+    } else if (userEvent.active != null) {
```
Comment (sohanirao) [inline src/runtime/propensity.js]: Now that we are not enforcing userEvent.active to be null and not undefined in case the caller has no information about the event being from a user action or not, we need to update your Event Manager code to not fail validation of the event and corresponding tests. You can either make that change first as a different PR or roll that into this one.

Comment (sohanirao) [reply src/runtime/propensity.js]: I looked into this. Your point is valid. We need to send something, undefined won't do. However, I wanted to find a way to handle it that is consistent with the rest of our code based. I found several instances of '|| null' to cover cases where undefined could be passed into an API for the field of an object defined using typedef.

Whenever a typedef requires something to be set and won't accept undefined, we OR the incoming object with a null. In this case it would look like this.

    this.eventManager_.logEvent({
    this.propensityServer_.sendEvent(userEvent.name, paramString);	      eventType: analyticsEvent,
      eventOriginator: EventOriginator.PROPENSITY_CLIENT,
      isFromUserAction: userEvent.active || null,
      additionalParameters: data,
    });

I still think we don't have to have a hard check to ensure its not undefined because the code already doesn't allow it. So != is sufficient IMHO. But to your point, to make it comply, following the code pattern in the rest of the code and ORing with null should address your concerns. 
 Let me know your thoughts.

Comment (sohanirao) [reply src/runtime/propensity.js]: given that we have a new way to handle undefined, can we update the message on 101 to be Active field must be boolean. 
Because if its not boolean, both null or undefined are okay. we handle it in line 107.

Comment (mborof) [reply src/runtime/propensity.js]: I'm still quite convinced that only test code can ever pass in undefined because the compiler will reject anything that allows undefined.  The reason it works in the tests is that the compiler seems to be less strict for them.  Real code won't be able to do it (except perhaps through Propensity where the publishers might not be doing strict compilation checking).  

I'd also strongly prefer not allowing it to be either undefined or null and treating them as the same, but it isn't actually all that important so I've just gone ahead and made that change.

Comment (mborof) [reply src/runtime/propensity.js]: I didn't see the previous 2 comments.  I don't think the || null is necessary because Propensity.Event doesn't support not having active.

Comment (sohanirao) [reply src/runtime/propensity.js]: Believe we have a resolution to this.
